Implement waiting for Connect My Computer node to join cluster

ravicious opened the pull request:
https://github.com/gravitational/teleport/blob/master/rfd/0133-connect-my-computer.md#determining-a-successful-launch

In short, after we start a Teleport agent in Connect, we want to wait until the node successfully joins the cluster so that we can change the UI state from "starting" to "running".

This is implemented by an RPC in tsh daemon which reads host UUID file from disk and then sets up a watcher to wait for `OpPut` of the node. Most of the work is done by `NodeJoinWait` in the `connectmycomputer` package. When the Electron app wants to wait until the node join, it calls this RPC.

Best reviewed commit-by-commit.

---

If you'd like to test it on a real cluster, then you need to checkout `ravicious/launch-js`, [open the config file](https://goteleport.com/docs/connect-your-client/teleport-connect/#configuration), set `feature.connectMyComputer` to `true`, restart the app and then click the laptop icon in the upper right of the view of all cluster resources.

The setup and subsequent launches should show a success state only once we detect that the node has joined the cluster.

ravicious committed:
Add daemon.Service.ResolveClusterURI

ravicious committed:
Implement WaitForConnectMyComputerNodeJoin in tsh daemon

ravicious committed:
Accept agents dir through command line flag

tshd needs to know this out of band, so that when the Electron app tells
it to watch for host UUID file for a specific cluster, the Electron app
can send just the profile name of the cluster instead of an arbitrary path
on the computer.

ravicious committed:
Implement WaitForConnectMyComputerNodeJoin in Electron app

ravicious commented on lib/teleterm/services/connectmycomputer/connectmycomputer.go:
```diff
@@ -354,3 +329,167 @@ type Provisioner interface {
+	}
+	defer watcher.Close()
+
+	// Attempt to fetch the node from the cluster manually, in case it's joined the cluster before we
+	// started the watcher.
+	//
+	// This means that we might return immediately if the node is still in the cache, even if
+	// technically the agent has not joined the cluster yet. We're fine with this edge case.
+	server, err := accessAndIdentity.GetNode(ctx, defaults.Namespace, nodeName)
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return nil, trace.Wrap(err)
+		}
+		// Continue in case of NotFound error.
+	} else {
+		return server, nil
+	}
+
+	resource, err := waitForOpPut(ctx, watcher, types.KindNode, nodeName)
```
Is there an easy way to parallelize both `GetNode` and `waitForOpPut` so that instead of waiting for a response from `GetNode` before processing events, we do both at the same time and return whatever is faster?

I wanted to use goroutines and a buffered channel to send the server over it, but I don't know how to handle errors when channels come into play. I'd like the semantics around errors to remain the same.

ravicious replied:
I won't be parallelizing this in the end. The cost of not doing this is very small. And from what I've seen on Slack, any attempt at making it so would make the code overall much more complex.

ravicious commented on lib/teleterm/services/connectmycomputer/connectmycomputer.go:
```diff
@@ -324,6 +330,119 @@ type Provisioner interface {
+	for {
+		select {
+		case <-ticker.C:
+			out, err := utils.ReadPath(hostUUIDPath)
+			if err != nil {
+				if trace.IsNotFound(err) {
+					continue
+				}
+				return "", trace.Wrap(err)
+			}
+
+			id := strings.TrimSpace(string(out))
+			if id == "" {
+				return "", trace.NotFound("host UUID is empty")
+			}
+
+			return id, nil
```
I couldn't use `utils.ReadHostUUID` because it returns `NotFound` both for when the file doesn't exist and when the host UUID in the file is empty.

I need to be able to distinguish between those two cases to enter the for loop again if the file was not created yet.

ibeckermayer replied:
Worth an inline comment

ravicious committed:
wait: Use addEventListener instead of onabort

ravicious committed:
Make TshAbortController emit abort event only once

This aligns it with a regular AbortController, which also emits the event
only once.

ravicious committed:
Refactor how types are imported in tshd fixtures

ibeckermayer commented on integration/teleterm_test.go:
```diff
@@ -678,6 +686,63 @@ func testCreatingAndDeletingConnectMyComputerToken(t *testing.T, pack *dbhelpers
+	)
+	require.NoError(t, err)
+
+	profileName, _, err := net.SplitHostPort(pack.Root.Cluster.Web)
```
unchecked error

ibeckermayer commented on lib/teleterm/services/connectmycomputer/connectmycomputer.go:
```diff
@@ -324,6 +330,119 @@ type Provisioner interface {
+}
+
+func (n *NodeJoinWait) getNodeNameFromHostUUIDFile(ctx context.Context, cluster *clusters.Cluster) (string, error) {
+	hostUUIDPath := filepath.Join(n.cfg.AgentsDir, cluster.ProfileName, "data", utils.HostUUIDFile)
```
Is there some way to create this path that ensures it stays consistent with whatever piece of code creates the file, which presumably needs to construct the same path?

ibeckermayer replied:
I figured that the file was being created somewhere with Go, and so was wondering if there might be some way to marry the code, however I see that that's not the case. Your refactor seems fine nevertheless.

ibeckermayer commented on lib/teleterm/services/connectmycomputer/connectmycomputer.go:
```diff
@@ -324,6 +330,119 @@ type Provisioner interface {
+	ticker := time.NewTicker(50 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
```
Should we have some sort of timeout here so we don't loop endlessly in the case that there's a problem creating the file?

ravicious replied:
To address both of your comments about timeouts, I didn't add any in the Go code because I expect the client to set for the whole request. The frontend app is going to abort this request after 20 seconds.

However, as you said, both operations essentially block forever if an error happens somewhere else. I suppose a timeout is warranted after all because of that, so I'm going to add a 1 minute timeout to `NodeJoinWait.Run`.

ibeckermayer commented on lib/teleterm/services/connectmycomputer/connectmycomputer.go:
```diff
@@ -324,6 +330,119 @@ type Provisioner interface {
+		return server, nil
+	}
+
+	resource, err := waitForOpPut(ctx, watcher, types.KindNode, nodeName)
```
Wrong commit to comment, but I'm not seeing where we set a timeout for `waitForOpPut`.

gzdunek commented:
For some reason, after completing the setup, `WaitForConnectMyComputerNodeJoin` doesn't return the hostname value in labels:
```
[tshd] info: receive: /teleport.lib.teleterm.v1.TerminalService/WaitForConnectMyComputerNodeJoin -> ({"server":{"uri":"/clusters/mercury.cloud.gravitational.io/servers/fd4c100e-47d4-489d-a0f0-edfcc9eaf086","tunnel":true,"name":"fd4c100e-47d4-489d-a0f0-edfcc9eaf086","hostname":"mbp.home","addr":"","labelsList":{"0":{"name":"hostname","value":""},"1":{"name":"teleport.dev/connect-my-computer/owner","value":"grzegorz.zdunek@goteleport.com"}}}})
```

However, on subsequent calls it works. 

Does it happen to you too?

ravicious commented:
@gzdunek That's probably because the hostname label is dynamic and it's not yet set when the first `OpPut` event comes in. On subsequent calls, the RPC returns the node through `GetNode` rather than from the event.

I suspected that returning the resource from the event might bite me, but not in this way.

We could certainly attempt to refetch the node after receiving the event rather than returning it straight from the event. What do you think?

gzdunek commented:
>We could certainly attempt to refetch the node after receiving the event rather than returning it straight from the event. What do you think?

Yeah, let's do this.

ibeckermayer commented on lib/teleterm/services/connectmycomputer/connectmycomputer.go:
```diff
@@ -348,12 +348,21 @@ func NewNodeJoinWait(cfg *NodeJoinWaitConfig) (*NodeJoinWait, error) {
 // The Electron app calls this method soon after starting the agent process.
 func (n *NodeJoinWait) Run(ctx context.Context, accessAndIdentity AccessAndIdentity, cluster *clusters.Cluster) (clusters.Server, error) {
-	nodeName, err := n.getNodeNameFromHostUUIDFile(ctx, cluster)
+	// The Electron app aborts the request which calls NodeJoinWait after a timeout, but let's use a
+	// timeout internally as well. Both operations in this struct theoretically block forever if an
+	// error happens elsewhere and gRPC doesn't set a timeout by default.
+	//
+	// If there's a bug in the Electron app and it forgets to abort the request, at least the request
+	// will not hang forever.
```
```suggestion
	// The Electron app aborts this request after a timeout, but we set a timeout here as well for
	// to ensure that the request can't hang forever if there's a bug in the request abort logic.
```

Two things to consider:

1. Perhaps move this `timeoutCtx` up into [the endpoint itself](https://github.com/gravitational/teleport/blob/e384604811d9c2f73231efd0a16133d5dbcb993d/lib/teleterm/apiserver/handler/handler_connectmycomputer.go#L59-L73). That way it's clear that the whole endpoint just times out after a minute.
2. I'm not sure how big of a change this would be, but since we're adding timeout logic here anyways, there may be a case for just removing it from the frontend code and using this logic exclusively instead, or having the frontend pass the timeout as a request parameter.

ravicious replied:
> 1. Perhaps move this `timeoutCtx` up into [the endpoint itself](https://github.com/gravitational/teleport/blob/e384604811d9c2f73231efd0a16133d5dbcb993d/lib/teleterm/apiserver/handler/handler_connectmycomputer.go#L59-L73). That way it's clear that the whole endpoint just times out after a minute.

Yeah, I wasn't sure if it'd be best to add a timeout tightly around the operation that actually blocks vs for the whole endpoint.

Since those individual services technically could serve as building blocks for other requests, I guess it's best to add the timeout on the endpoint level.

> 2\. I'm not sure how big of a change this would be, but since we're adding timeout logic here anyways, there may be a case for just removing it from the frontend code and using this logic exclusively instead, or having the frontend pass the timeout as a request parameter.

Now that I think about it, under normal circumstances this would be the way to go. However, the UI needs to be able to distinguish between a timeout and other errors. The JS gRPC client encodes that information as custom properties on the error object. gRPC errors in our app pass through [the Electron context bridge](https://www.electronjs.org/docs/latest/api/context-bridge) which strips all custom properties from Error objects, meaning we'd only have the error message to distinguish between a timeout or not.

So until we revamp how those errors are passed, I think I'm going to keep current implementation as is with regards to the timeout.

I'm still not sure if an endpoint should trouble itself with the deadline, perhaps it should rather return `BadParameter` if no deadline was set on the gRPC request? I'd rather use built-in mechanisms to handle stuff like this vs building my own on top of gRPC.

ravicious commented on lib/teleterm/services/connectmycomputer/connectmycomputer.go:
```diff
@@ -354,3 +329,172 @@ type Provisioner interface {
+		return server, nil
+	}
+
+	resource, err := waitForOpPut(ctx, watcher, types.KindNode, nodeName)
```
@gzdunek Regarding the problem with a blank hostname, it's not going to be as simple as refetching the node after `OpPut` comes in.

I just implemented this and even when connecting to a cluster in our cloud, the label update doesn't happen fast enough. We could wait for the hostname label to get filled out, but this would complicate the implementation and I don't think it's entirely necessary.

What if we used some smoke and mirrors here? If there's a hostname label (and there will be one by default) and it's empty, we fill it out with what `os.Hostname` reports.

gzdunek replied:
>We could wait for the hostname label to get filled out, but this would complicate the implementation and I don't think it's entirely necessary.

Yeah, it is not worth it.

>If there's a hostname label (and there will be one by default) and it's empty, we fill it out with what os.Hostname reports.

Works for me!

ravicious committed:
createAbortController: Add signal.aborted, use emitter.once

ravicious committed:
Improve wait function based on Deno implementation

https://github.com/denoland/deno_std/blob/72d6e6641e3cd39ae69fba89a78feab354018ef0/async/delay.ts#L39

ravicious committed:
Add a comment about the events package

ravicious commented:
Ping @ibeckermayer, we have two other PRs stacked on this one so I'd like to merge it if possible. ;)

ibeckermayer commented on lib/teleterm/services/connectmycomputer/connectmycomputer.go:
```diff
@@ -358,6 +360,14 @@ func (n *NodeJoinWait) Run(ctx context.Context, accessAndIdentity AccessAndIdent
 		return clusters.Server{}, trace.Wrap(err)
 	}
 
+	// By default, the node config includes a command label with a hostname. Immediately after joining
+	// the cluster, the label is most likely not going to be updated. To work around that, we fill it
+	// out with os.Hostname if it's empty.
```
What this comment is pointing out isn't quite clear to me (granted it could just be a function of the fact that I'm unfamiliar with this part of the codebase).

ravicious replied:
Oh, I think the comment doesn't make it clear that the config comes from `teleport node config` that we call during the setup of Connect My Computer.

I'll update the comment to my best ability and then merge this PR, but please let me know if the updated comment is any better.

ravicious commented:
Merged #30910 into this PR with fast forward.
